cmd/mutate-image/main.go

```diff
@@ -0,0 +1,259 @@
+	pflag.StringVar(&flagValues.resultFileImageSize, "result-file-image-size", "", "A file to write the image size to")
+}
+
+func main() {
+	if err := Execute(context.Background()); err != nil {
+		exitcode := 1
+
+		switch err := err.(type) {
+		case *ExitError:
+			exitcode = err.Code
+		}
+
+		log.Print(err.Error())
+		os.Exit(exitcode)
+	}
+}
+
+// Execute performs flag parsing, input validation and the image mutation
+func Execute(ctx context.Context) error {
+	flagValues = settings{}
+	pflag.Parse()
+	return runMutateImage(ctx)
+}
+
+func runMutateImage(ctx context.Context) error {
+	if flagValues.image == "" {
+		return &ExitError{Code: 100, Message: "the 'image' argument must not be empty"}
+	}
+
+	options := getOptions(ctx)
+	ref := flagValues.image
+
+	if len(flagValues.annotation) != 0 {
+		desc, err := crane.Head(ref, *options...)
+		if err != nil {
+			return fmt.Errorf("checking %s: %v", ref, err)
+		}
+
+		if desc.MediaType.IsIndex() {
+			return fmt.Errorf("mutating annotations on an index is not yet supported")
+		}
+	}
+
+	img, err := crane.Pull(ref, *options...)
+	if err != nil {
+		return fmt.Errorf("pulling %s: %v", ref, err)
+	}
+
+	cfg, err := img.ConfigFile()
+	if err != nil {
+		return fmt.Errorf("getting config: %v", err)
+	}
+	cfg = cfg.DeepCopy()
+
+	// Set labels.
+	if cfg.Config.Labels == nil {
+		cfg.Config.Labels = map[string]string{}
+	}
+
+	labels, err := splitKeyVals(flagValues.label)
+	if err != nil {
+		return err
+	}
+
+	for k, v := range labels {
+		cfg.Config.Labels[k] = v
+	}
+
+	annotations, err := splitKeyVals(flagValues.annotation)
+	if err != nil {
+		return err
+	}
+
+	// Mutate and write image.
+	img, err = mutate.Config(img, cfg.Config)
+	if err != nil {
+		return fmt.Errorf("mutating config: %v", err)
+	}
+
+	img = mutate.Annotations(img, annotations).(v1.Image)
+
+	digest, err := img.Digest()
+	if err != nil {
+		return fmt.Errorf("digesting new image: %v", err)
+	}
+
+	r, err := name.ParseReference(ref)
+	if err != nil {
+		return fmt.Errorf("parsing %s: %v", ref, err)
+	}
+
+	if _, ok := r.(name.Digest); ok {
+		ref = r.Context().Digest(digest.String()).String()
+	}
+
+	if err := crane.Push(img, ref, *options...); err != nil {
+		return fmt.Errorf("pushing %s: %v", ref, err)
+	}
+
+	fmt.Println(r.Context().Digest(digest.String()))
+
+	// Writing image digest to file
+	if resultFileImageDigest := flagValues.resultFileImageDigest; resultFileImageDigest != "" {
+		if err := ioutil.WriteFile(
+			resultFileImageDigest, []byte(digest.String()), 0644,
+		); err != nil {
+			return err
+		}
+	}
+
+	// Writing image size in bytes to file
+	if resultFileImageSize := flagValues.resultFileImageSize; resultFileImageSize != "" {
+		size, err := GetCompressedImageSize(img)
+		if err != nil {
+			return err
+		}
+
+		if err := ioutil.WriteFile(
+			resultFileImageSize, []byte(strconv.FormatInt(size, 10)), 0644,
+		); err != nil {
+			return err
+		}
+	}
```

Comment:
Interesting, I had not expected we would write the image size here because annotations and labels should not change it (at least I thought we only modify the manifest which is not counted in the image size). Can you run a manual verification that the result of a build that supports the image size (like [kaniko](https://github.com/shipwright-io/build/blob/main/samples/buildstrategy/kaniko/buildstrategy_kaniko_cr.yaml#L64)) produces the same size without and with annotations. If it works I'd say we can leave it in.

Reply:
I have verified the change in image compressed size by adding the annotations and labels, annotations won't make any changes in compressed size but the labels do.